CODING_STYLE.md

```diff
@@ -0,0 +1,460 @@
+  ```
+
+#### Return Error Codes
+
+- When writing a function that can fail in a non-binary way try to return
+  meaningful negative error codes (e.g. `return -EINVAL;`).
+
+#### All Unexported Functions Must Be Declared `static`
+
+- Functions which are only used in the current file and are not exported
+  within the codebase need to be declared with the `static` attribute.
+
+#### All Exported Functions Must Be Declared `extern` In A Header File
+
+- Functions which are used in different files in the library should be declared
+  in a suitable `*.c` file and exposed in a suitable `*.h` file. When defining
+  the function in the `*.c` file the function signature should not be preceded
+  by the `extern` keyword. When declaring the function signature in the `*.h`
+  file it must be preceded by the `extern` keyword. For example:
+  ```
+  /* Valid function definition in a *.c file */
+  ssize_t lxc_write_nointr(int fd, const void* buf, size_t count)
+  {
+          ssize_t ret;
+  again:
+          ret = write(fd, buf, count);
+          if (ret < 0 && errno == EINTR)
+                  goto again;
+          return ret;
+  }
+
+  /* Valid function declaration in a *.h file */
+  extern ssize_t lxc_write_nointr(int fd, const void* buf, size_t count);
+  ```
+
+#### All Names Must Be In lower_case
+
+- All functions and variable names must use lower case.
+
+#### Declaring Variables
+
+- variables should be declared at the top of the function or at the beginning
+  of a new scope but **never** in the middle of a scope
+1. uninitialized variables
+  - put base types before complex types
+  - put standard types defined by libc before types defined by LXC
+  - put multiple declarations of the same type on the same line
+2. initialized variables
+  - put base types before complex types
+  - put standard types defined by libc before types defined by LXC
+  - put multiple declarations of the same type on the same line
+- Examples of good declarations can be seen in the following function:
+  ```
+  int lxc_clear_procs(struct lxc_conf *c, const char *key)
+  {
+          struct lxc_list *it, *next;
+          bool all = false;
+          const char *k = NULL;
+
+          if (strcmp(key, "lxc.proc") == 0)
+                  all = true;
+          else if (strncmp(key, "lxc.proc.", sizeof("lxc.proc.") - 1) == 0)
+                  k = key + sizeof("lxc.proc.") - 1;
+          else
+                  return -1;
+
+          lxc_list_for_each_safe(it, &c->procs, next) {
+                  struct lxc_proc *proc = it->elem;
+
+                  if (!all && strcmp(proc->filename, k) != 0)
+                          continue;
+                  lxc_list_del(it);
+                  free(proc->filename);
+                  free(proc->value);
+                  free(proc);
+                  free(it);
+          }
+
+          return 0;
+  }
+    ```
+
+#### Single-line `if` blocks should not be enclosed in `{}`
```

Comment:
I always find this one to be potentially dangerous :)

Reply:
Not really, the compiler will yell at you if you have a dangling else or something. It's just way easier on the eye and we've always been doing it this way. :)

Reply:
The compiler will yell at me if I do?

```c
if (1 == 0)
    system("rm -Rf /");
    system("rm -Rf /");
```

Reply:
```
stgraber@castiana:~$ cat test.c
#include <stdio.h>

void main() {
    if (1 == 0)
        printf("Don't run this\n");
        printf("Or this\n");
}
stgraber@castiana:~$ gcc test.c -o test
stgraber@castiana:~$ ./test
Or this
```

Compiler doesn't seem to be yelling at me that I'm probably missing some brackets here.

Reply:
No, my point was about:

if (bla)
    if (blu)
  else

That was unclear. There's nothing to prevent you from doing the nonsense
you just outlined
but coding style wise this isn't really valid: clang-analyzer will correct
the indentation at which
point the error should be obvious. Same for GNU indent and other code.

Reply:
Would also be interesting what `gcc` and `clang` have to say with `-Wmisleading-indentation` btw.

Reply:
Yup, they will yell at you:

```
chb@conventiont|~/files/Cground|github-master $%=
> gcc -Wmisleading-indentation -o test test.c
test.c: In function ‘main’:
test.c:307:5: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
     if (1 == 0)
     ^~
test.c:309:9: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
         printf("Or this\n");
         ^~~~~~
```